source/adapters/level_zero/v2/command_buffer.cpp

```diff
@@ -0,0 +1,309 @@
+  const char *UrRet = std::getenv(AppendEnvVarName);
+
+  if (!Device->ImmCommandListUsed) {
+    logger::error("Adapter v2 is used but immediate command-lists are currently "
+                  "disabled. Immediate command-lists are "
+                  "required to use the adapter v2.");
+    std::abort();
+  }
+  if (!DriverSupportsImmediateAppend) {
+    logger::error("Adapter v2 is used but "
+                  "the current driver does not support the "
+                  "zeCommandListImmediateAppendCommandListsExp entrypoint. A "
+                  "driver version of at least {} is required to use the "
+                  "immediate append path.", MinDriverVersion);
+    std::abort();
+  }
+
+  const bool EnableAppendPath = !UrRet || std::atoi(UrRet) == 1;
+  if (!Device->isPVC() && !EnableAppendPath) {
+    logger::error("Adapter v2 is used but "
+                  "immediate append support is not enabled."
+                  "Please set {}=1 to enable it.", AppendEnvVarName);
+    std::abort();
+  }
+
+}
+
+}
+
+std::pair<ze_event_handle_t *, uint32_t>
+ur_exp_command_buffer_handle_t_::getWaitListView(
+    const ur_event_handle_t *phWaitEvents, uint32_t numWaitEvents) {
+
+  waitList.resize(numWaitEvents);
+  for (uint32_t i = 0; i < numWaitEvents; i++) {
+    waitList[i] = phWaitEvents[i]->getZeEvent();
+  }
+
+  return {waitList.data(), static_cast<uint32_t>(numWaitEvents)};
+}
+
+ur_exp_command_buffer_handle_t_::ur_exp_command_buffer_handle_t_(
+    ur_context_handle_t Context, ur_device_handle_t Device,
+      ze_command_list_handle_t CommandList,
+      const ur_exp_command_buffer_desc_t *Desc)
+    : Context(Context), Device(Device), ZeCommandList(CommandList),
+      IsUpdatable(Desc ? Desc->isUpdatable : false) {
+  UR_CALL_THROWS(ur::level_zero::urContextRetain(Context));
+  UR_CALL_THROWS(ur::level_zero::urDeviceRetain(Device));
+}
+
+void ur_exp_command_buffer_handle_t_::cleanupCommandBufferResources() {
+  // Release the memory allocated to the Context stored in the command_buffer
+  UR_CALL_THROWS(ur::level_zero::urContextRelease(Context));
+
+  // Release the device
+  UR_CALL_THROWS(ur::level_zero::urDeviceRelease(Device));
+
+  // Release the memory allocated to the CommandList stored in the
+  // command_buffer
+  if (ZeCommandList) {
+    ZE_CALL_NOCHECK(zeCommandListDestroy, (ZeCommandList));
+  }
+
+  for (auto &AssociatedKernel : KernelsList) {
+    UR_CALL_THROWS(ur::level_zero::urKernelRelease(AssociatedKernel));
+  }
+}
+
+namespace ur::level_zero {
+
+/**
+ * Creates a L0 command list
+ * @param[in] Context The Context associated with the command-list
+ * @param[in] Device  The Device associated with the command-list
+ * @param[in] IsUpdatable Whether the command-list should be mutable.
+ * @param[out] CommandList The L0 command-list created by this function.
+ * @return UR_RESULT_SUCCESS or an error code on failure
+ */
+ur_result_t createMainCommandList(ur_context_handle_t Context,
```

Comment:
Would it make sense to cache command lists like we do for non-command-buffer path?

Reply:
I don't know what the non-command-buffer path does, but for reference creating a pool of command-lists to use has been an idea we've had for the v1 adapter but never got around to. [TODO comment that has since been removed](https://github.com/intel/llvm/pull/9992/commits/c385466a68e5192ad38d5bfd4a4c2a6dd36ac4c7#r1247439851)

Reply:
In v2, we have https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/v2/command_list_cache.hpp which could be used here.